# Arabidopsis case study

## Build the tree and collect the proteomes

```{r, message=FALSE}
library(reshape2)
library(taxizedb)
library(dplyr)
library(readr)
library(phylostratr)
library(magrittr)
library(onekp)
library(plotly)
library(knitr)
```

```{r at_prep, cache=TRUE}
# First get core uniprot proteomes 
weights=uniprot_weight_by_ref()
focal_taxid <- '3702'
strata <-
  # Get stratified relatives represented in UniProt
  uniprot_strata(focal_taxid, from=2L) %>%
  # Select a diverse subset of 5 or fewer representatives from each stratum.
  strata_apply(f=diverse_subtree, n=5L, weights=weights) %>%
  # Use a prebuilt set of prokaryotic species
  use_recommended_prokaryotes %>%
  add_taxa(c('4932', '9606')) %>%
  # Download genomes, storing the filenames
  uniprot_fill_strata

# Next add a few supplementary proteomes from 1KP
kp <- onekp::retrieve_onekp()
# get the ginkgo proteome from 1KP
ginkgo_taxid <- taxizedb::name2taxid("Ginkgo biloba")
ginkgo_proteome <- onekp::filter_by_species(kp, ginkgo_taxid) %>% onekp::download_peptides(dir="custom-seqs")
strata <- add_taxa(strata, ginkgo_taxid)
strata@data$faa[[as.character(ginkgo_taxid)]] <- as.vector(ginkgo_proteome)

# Remove Klebsormidium nitens
strata <- prune(strata, '105231', type='name')

## Replace the UniProt Arabidopsis thaliana proteome with the Araport11 proteome.
## The path below is specific to my pipeline, you can just use the UniProt genes,
## if you want. The problem with them is that they include all previous variants
## of the gene models, raising to 89135 genes. Since `A. thaliana` is the focal
## species, it is important to be very explicit about version of the proteome.
strata@data$faa[['3702']] <- 'Araport11_genes.201606.pep.fasta'
```

## BLAST focal species against all target species

The next step is to BLAST the *A. thaliana* proteins against each target
proteome.

This will probably take a long time to finish. We strongly recommend running it
on a cluster with the search against each target proteome as a separate job or
using [GNU Parallel](http://www.gnu.org/software/parallel) to speed up the
process ([see
guide](https://gif.biotech.iastate.edu/running-blast-jobs-parallel)).
`phylostratr` expects one tabular file of BLAST results for each comparison
against each species. The files should head headers and should contain the
following fields: `qseqid`, `sseqid`, `qstart`, `qend`, `sstart`, `send`,
`evalue`, `score` The name of the BLAST result files should have the form
`<target_name>.tab`, where `<target_name>` is the name stored in the
`phylostratr` `Strata` object.

While the details will depend on your needs and the particulars of your system,
for each target you will want something of this general form:

``` sh
echo -e 'qseqid\tsseqid\tqstart\tqend\tsstart\tsend\tevalue\tscore' > <target>.tab
blastp \
    -query <focal_species> \
    -db <target_database> \
    -outfmt '6 qseqid sseqid qstart qend sstart send evalue score' >> <target>.tab
```

Then move the resulting files into the working directory where you are running
`phylostratr`. Then you may proceed with the following code (if everything went
well, it will automatically read these BLAST results).


```{r, at_blast, cache=TRUE, message=FALSE, warning=FALSE}
# replace 8 with the number of threads on your system, if you running locally
strata <- strata_blast(strata, blast_args=list(nthreads=8L)) %>%
  strata_besthits

# Merge results into a single hittable
results <- merge_besthits(strata)

# Get metadata. Note, this will take awhile
strata <- strata %>%
    # for each species, add map of UniProt IDs to PFAM domains
    strata_uniprot_pfam_map %>%
    # for each species, add proteome summary data
    add_proteome_stats %>%
    # for each species, add list of organelle encoded UniProt proteins
    add_organelle_proteins
```

## Proteome stats tables

```{r, at_proteome_stats, cache=TRUE}
prot <- proteome_stats_table(strata)
strata2 <- strata_convert(strata, target='all', to='name')
```

```{r at_prot_stat, eval=FALSE}
# You can explore the proteome stats interactively with plotly:
plotly::ggplotly(plot_proteome_stats(strata2))
plotly::ggplotly(plot_proteome_lengths(strata2))
```

```{r, at_protreport, cache=FALSE}
proteome_report_table(strata2) %>%
  merge(get_phylostrata_map(strata2)) %>%
  dplyr::arrange(-ps) %>%
  dplyr::select(species, ps, N, min, q25, median, q75, max) %>%
  knitr::kable()
```

Table: Proteome statistics for all proteomes in the *A. thaliana* analysis. **N**: total
number of proteins in the UniProt proteome (these may be redundant). **ps**:
phylostratum level, where 1 is *cellular organisms*. **min**, **q25**,
**median**, **q75**, **max**: summaries of the protein lengths in each
proteome. 

## Identification of organelle proteins 


```{r}
# Select just the plant species 
subtree(strata, name2taxid('Viridiplantae')) %>%
    # replace the taxon IDs with scientific names
    strata_convert(target='all', to='name') %>%
    # summarize the numbers of organelle genes present per species
    organelle_table %>%
    # sort by ascending phylostratum age
    dplyr::arrange(-.data$ps, .data$species) %>%
    knitr::kable()
```

Table: Number of organelle-encoded proteins appearing in each plant target
species. Currently, organelle proteins can only be counted for sequeces
retrieved from UniProt (which have UniProt IDs in their FASTA headers). Thus
*A. thaliana* and *Ginkgo biloba* are each listed as having no organellar
sequences.


## Finding genes that skip strata

```{r, cache=TRUE}
# Find all genes that are missing in one or more phylostrata younger than their
# oldest match.
revenants <- find_revenants(results)
```

```{r, cache=TRUE}
# The number of genes with a skipped stratum immediately after
# the basal stratum
(subset(revenants, ps - 1 == basal_ps)$qseqid %>% unique %>% length)
```

```{r, cache=TRUE}
# Summarize of the number of genes that skip $n\_skipped$ strata.
revenants %>%
    dplyr::distinct(qseqid, n) %>%
    dplyr::group_by(n) %>%
    dplyr::count() %>%
    dplyr::select(n_skipped=n, n_total=nn) %>%
    knitr::kable()
```

Table: A summary of the number of *A. thaliana* genes (**n_total**) where a
given number (**n_skipped**) of strata are passed before the oldest stratum
with a predicted homolog is reached. 

```{r, cache=TRUE}
# Show the skipped phylostrata for genes skipping more than 4 
revenants %>%
    dplyr::filter(n > 4) %>%
    dplyr::select(qseqid, ps, basal_ps, n) %>%
    dplyr::arrange(-ps) %>%
    dplyr::group_by(qseqid) %>%
    dplyr::summarize(basal_ps=basal_ps[1], skipped_ps = paste(ps, collapse=",")) %>%
    dplyr::arrange(-basal_ps) %>%
    knitr::kable()
```

Table: Here we identify which strata were skipped for each of the *A. thaliana*
genes that skipped more than 4 strata.

```{r, cache=TRUE}
revenants %>%
    dplyr::group_by(mrca) %>%
    dplyr::summarize(missing_genes = n(), ps=ps[1]) %>%
    dplyr::arrange(-ps) %>%
    dplyr::mutate(mrca = partial_id_to_name(mrca)) %>%
    dplyr::select(Phylostratum = mrca, Level = ps, missing_genes) %>%
    knitr::kable()
```

Table: The number of *A. thaliana* genes that have no hits to any species in a
given phylostratum but do have hits to older strata. Strata may have
unexpectedly high number of "missing genes" for interesting biological reasons,
for example, the branch may include degenerate intracellular parasites. But
more likely, it is due to poor annotations of the stratum representatives; this
is the case for the Spermatophyta stratum, which is represented only by *Picea
glauca*. 

## Compare to phylostratigraphy results from 2014 

Here I will load the phylostratigraphy results from (Arendsee et al., 2014) and
compare them to the new results from the `phylostratr` algorithm. The
algorithms and cutoffs are the same, but the search space is different. Also,
in the new study, I use the Araport11 anotation of *Arabidopsis thaliana*,
whereas in (Arendsee et al., 2014) TAIR10 was used.

```{r, at_tips_prep, cache=TRUE, message=FALSE}
# Get the strata from the current analysis
a2018 <- stratify(results)
# Get and clean the (Arendsee et al., 2014) data (stored in phylostratr package)
a2014 <- read_tsv(system.file('extdata', 'arendsee2014_strata.tab', package='phylostratr'))
# Add NCBI taxonomy ID to tips data
a2014$mrca <- taxizedb::name2taxid(a2014$name)
# match column naming conventions
a2014 <- dplyr::select(a2014, qseqid, mrca, ps, mrca_name=name)
# underscores to spaces 
a2014$mrca_name <- sub('_', ' ', a2014$mrca_name)
# get a lineage map for Arabidopsis
strata_map <- taxizedb::classification('3702')[[1]]
# factor the backbone
a2014$mrca_name <- droplevels(factor(a2014$mrca_name, levels=strata_map$name))
```

Now, since the two phylostratigraphy experiments used different annotations of
Arabidopsis, I will take the intersection. The model IDs shared between the
annotations refer to the same models, phylostratr has a dedicated
function for this:

```{r, at_tips_summary, cache=TRUE}
ps <- standardize_strata(list(
    a2014 = a2014,
    a2018 = a2018
))

# Compare the counts between the two studies:
all_strata <- do.call(rbind, ps) %>%
    merge(strata_map, by.x='mrca', by.y='id') %>%
    dplyr::select(qseqid=qseqid, mrca=mrca, mrca_name=name, group=group) %>%
    dplyr::mutate(mrca_name = droplevels(factor(mrca_name, levels=strata_map$name)))

# Compare the values between the two studies
strata_summary <- all_strata %>%
    dplyr::group_by(group, mrca_name) %>%
    dplyr::summarize(count = n())

# make the table
dcast(strata_summary, mrca_name ~ group, value.var='count') %>%
    knitr::kable()
```

```{r, orphan_diff}
d <- subset(all_strata, mrca_name == "Arabidopsis thaliana")
a2014_orphans <- subset(d, group == "a2014")$qseqid
a2018_orphans <- subset(d, group == "a2018")$qseqid
setdiff(a2018_orphans, a2014_orphans)
```

Table: A comparison between (Arendsee et al., 2014, `a2014`) and this study's
(`a2018`) stratification of *A. thaliana* genes. 
