
# Yeast case study

First we collect UniProt proteomes for diverse representation across the phylostrata 

```{r, message=FALSE}
require(phylostratr)
require(magrittr)
require(dplyr)
require(readr)
require(knitr)
require(taxizedb)
require(ape)
```

```{r, cache=TRUE}
focal_taxid <- 4932
# set weights on species selection
weights <- c(
  '1355160' = 0,   # these three are deprecated
  '245562'  = 0,
  '1094981' = 0,
  '284813'  = 1.1, # these are the reference replacements
  '766039'  = 1.1,
  '237561'  = 1.1
)
focal_taxid <- '4932'
strata <- focal_taxid %>%
  # Get all UniProt proteomes
  uniprot_strata %>%
  # build a tree of all UniProt genomes
  # Select a diverse subset of 5 or fewer representatives from each stratum.
  # Only do this above the Saccharomyces genus, since we will later replace
  # Saccharomyces with out own tree.
  strata_apply(f=diverse_subtree, n=5, weights=weights) %>%
  # add prokaryote stratum
  use_recommended_prokaryotes %>%
  # download UniProt sequences (this may take 10+ minutes)
  uniprot_fill_strata
```

Next we replace the Saccharomyces genus with our own tree describing the
phylogenetic relationships between the species. We also provide our own
proteome sequence.

```{r, cache=TRUE}
saccharomyces <- Strata(
    tree = ape::read.tree(system.file('extdata', 'yeast', 'tree', package='phylostratr')),
    data = list(faa=list(
             # The user is responsible for collecting these data 
             Saccharomyces_cerevisiae   = 'yeast/cerevisiae.faa',
             Saccharomyces_paradoxus    = 'yeast/paradoxus.faa',
             Saccharomyces_mikatae      = 'yeast/mikatae.faa',
             Saccharomyces_kudriavzevii = 'yeast/kudriavzevii.faa',
             Saccharomyces_arboricola   = 'yeast/arboricola.faa',
             Saccharomyces_eubayanus    = 'yeast/eubayanus.faa',
             Saccharomyces_uvarum       = 'yeast/uvarum.faa'
           )),
    focal_species = 'Saccharomyces_cerevisiae'
) %>% strata_convert(target='tip', to='id')

strata <- replace_branch(strata, y=saccharomyces, node='4930') 
```

Now we can BLAST the focal species genome against the other species. This step
is exactly the same as the one in the Arabidopsis study.

```{r, sc_blast, cache=TRUE, message=FALSE, warning=FALSE}
strata  <- strata_blast(strata, blast_args=list(nthreads=2)) %>% strata_besthits
results <- merge_besthits(strata)
strata  <- strata %>%
    strata_uniprot_pfam_map %>%
    add_proteome_stats
```

## Proteome stats tables

```{r, at_proteome_stats}
strata2 <- strata_convert(strata, target='all', FUN=partial_id_to_name)
```

```{r sc_plot_stat, fig.width=8, fig.height=6, dpi=300}
plot_proteome_stats(strata2)
```

```{r sc_plot_length, fig.width=8, fig.height=6, dpi=300}
plot_proteome_lengths(strata2)
```

```{r, at_protreport, cache=FALSE}
proteome_report_table(strata2) %>%
  merge(get_phylostrata_map(strata2)) %>%
  dplyr::arrange(-ps) %>%
  dplyr::select(species, ps, N, min, q25, median, q75, max) %>%
  knitr::kable()
```
Table: Proteome statistics for all proteomes in the analysis. **N**: total
number of proteins in the UniProt proteome (these may be redundant). **ps**:
phylostratum level, where 1 is *cellular organisms*. **min**, **q25**,
**median**, **q75**, **max**: summaries of the protein lengths in each
proteome. 

## Finding genes that skip strata

Since the code is identical to that in the Arabidopsis study, it is elided here. 

```{r, echo=FALSE, cache=TRUE}
revenants <- find_revenants(results)
```

```{r, echo=FALSE, cache=TRUE}
revenants %>%
    dplyr::distinct(qseqid, n) %>%
    dplyr::group_by(n) %>%
    dplyr::count() %>%
    dplyr::select(n_skipped=n, n_total=nn) %>%
    knitr::kable()
```
Table: A summary of the number of genes (**n_total**) where a given number
(**n_skipped**) of strata are passed before the oldest stratum with a predicted
homolog is reached. 

```{r, echo=FALSE, cache=TRUE}
revenants %>%
    dplyr::filter(n > 4) %>%
    dplyr::select(qseqid, ps, basal_ps, n) %>%
    dplyr::arrange(-ps) %>%
    dplyr::group_by(qseqid) %>%
    dplyr::summarize(basal_ps=basal_ps[1], skipped_ps = paste(ps, collapse=",")) %>%
    dplyr::arrange(-basal_ps) %>%
    knitr::kable()
```
Table: Here we identify the strata that where skipped for all genes skipping
more than 4 strata. The genes with names other than `NM_XXXXXXXXX.X` are all
mitochondrial genes.


```{r, echo=FALSE, cache=TRUE}
revenants %>%
    dplyr::arrange(-ps) %>%
    dplyr::group_by(mrca) %>%
    dplyr::summarize(missing_genes = n()) %>%
    dplyr::mutate(mrca = partial_id_to_name(mrca)) %>%
    knitr::kable()
```
Table: We can break this dataset down to show the number of genes that have no
hits to any species in a given stratum (`mrca`) but do have hits to older
strata. In the table below, we see that `s2`, a stratum represented by *S.
arboricola*, has skipped far more often than the surrounding strata.


## Steps in building the tree 

```{r, cache=TRUE}
# just the order
sc_order <-
  focal_taxid %>%
  uniprot_strata %>%
  subtree(name2taxid('Saccharomycetales'))

# tree after filtering
sc_filtered <- sc_order %>% 
  strata_apply(f=diverse_subtree, n=5, weights=weights) %>%
  uniprot_fill_strata

# replace Saccharomyces branch with the users tree 
sc_final <- replace_branch(sc_filtered, saccharomyces, '4930')
```

```{r sc_tree1, dpi=300, fig.cap="Step 1: The NCBI common tree for all species in the Saccharomycetales family that are represented in UniProt proteomes"}
plot(sc_order@tree, show.node.label=TRUE, cex=0.45)
```

```{r sc_tree2, dpi=300, fig.cap="Step 2: The NCBI common tree after sampling a diverse subset of species"}
plot(sc_filtered@tree, show.node.label=TRUE, cex=0.45)
```

```{r sc_tree3, dpi=300, fig.cap="Step 3: The tree after replacing the Saccharomyces branch with a custom tree"}
plot(sc_final@tree, show.node.label=TRUE, cex=0.45)
```

# System Info

```{r}
# BLAST version
system2('blastp', '-version', stdout=TRUE)
```

```{r}
# R package versions and system info
require(devtools)
session_info()
```
