---
title: "*Arabidopsis thaliana* case study"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{*Arabidopsis thaliana* case study}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Select relatives to search

```{r, eval=FALSE}
library(phylostratr)
library(taxize)

focal_taxid <- 3702
strata <- focal_taxid %>%
  # Get stratified relatives represented in UniProt
  uniprot_cousins %>%
  # If there are more than 5 species in a stratum, take just the first from every uncle
  lapply(make_do_if_over(5, take_first)) %>%
  # Empty the ancestor of animals and fungi 
  clear_uncle(2759, 33154) %>%
  # Replace the representatives with humans and yeast
  add_representative(scinames=c('Homo sapiens', 'Saccharomyces cerevisiae')) %>%
  # Use a prebuilt set of prokaryotic species
  use_recommended_prokaryotes
```

```{r, eval=FALSE}
# Download genomes
strata_fasta <- uniprot_retrieve_proteomes(strata)

# Get a sample Arabidopsis thaliana protein FASTA file 
query  <- system.file(
    'extdata',
    'araport11',
    'at_ATxGx99xx.faa',
    package='phylostratr'
)

# add the focal species to the blast set
strata_fasta[['3702']] <- list('3702'=query)

# BLAST against each target genome (this will take a few hours)
results <- strata_blast(query, strata_fasta, blast_args=list(nthreads=8)) %>%
  strata_besthits %>%
  merge_besthits
```

The `results` variable is saved as a `phylostratr` dataset, and can be loaded
as below:

```{r, eval=FALSE}
data(arabidopsis)
head(arabidopsis)
```

You can get phylostrata:

```{r, eval=FALSE}
phylostrata <- stratify(arabidopsis)
head(phylostrata)
```

And get the number of genes in each strata


```{r, eval=FALSE}
summary(phylostrata$mrca_name)
```

Note that this is a reduced dataset, with only 1/100 of the genes represented.
