---
title: "*Arabidopsis thaliana* case study"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{*Arabidopsis thaliana* case study}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

`phylostratr` divides a phylostratigraphy analysis into 5 steps:

 1. Acquire a species tree with species identifiers (e.g. NCBI taxon ids or
    species names) as leafs.

 2. Replace leafs with the data required for homology inference. `phylostratr`
    supports retrieval of proteomes from UniProt.

 3. Run the homology inference algorithm (e.g. BLAST), producing a tree with
    raw results as leaves.

 4. From the raw results, produce p-values for homology of each species against
    each target.

 5. infer taxonomic specificity of each gene

Currently `phylostratr` supports using either custom trees or the NCBI common
tree. For homology inference, a simple protein BLAST is offered. However,
`phylostratr` is designed to be extendable to additional tools.

In this vignette, we will show how to stratify the genes of *Arabidopsis
thaliana* as follows:

 1. Trace the lineage of *Arabidopsis thaliana* as defined by the NCBI common
    tree. For each ancestral node, find species with proteomes available in
    UniProt. For over-represented strata, we select a diverse subset of species
    (for example, of the thousands of bacterial proteomes in uniprot, we select
    just one from each class). 

 2. Retrieve the proteome for each species in the tree, replacing the leaves
    with proteome file names. This allows us to access all the protein data
    needed for a protein BLAST (next step) while also storing the phylogenetic
    relationship between the species.

 3. BLAST the focal proteome (*Arabidopsis thaliana*) against each species in
    the tree, replacing leaves with the filenames of BLAST results.  We build a
    unique BLAST database for each species. Doing this makes the BLAST
    statistics against each individual species to be independent of the other
    species in the tree. Later we will calculate a cumulative p-value for
    existence of a homolog in a particular stratum.

 4. Next, for each BLAST result, find the best hit against each focal gene. The
    resulting tables (small enough to be easily stored in memory now) are
    recorded as the new leaves of the tree.

 5. Finally, the best hit tables are merged, and the genes are stratified. In
    this vignette, we use a simple classifier for homology: a subject species
    is inferred to have a homolog to a focal gene if BLAST reports an evalue of
    less than 1e-5. This is a fairly arbitrary criterion and does not attempt
    to adjust for multiple testing across a stratum.

## 1. Select relatives to search

```{r, eval=FALSE}
library(phylostratr)
library(taxize)

# TODO: use TOL instead of NCBI common tree

focal_taxid <- 3702
strata <- focal_taxid %>%
  # Get stratified relatives represented in UniProt
  uniprot_cousins %>%
  # If there are more than 5 species in a stratum, take just the first from every uncle
  lapply(make_do_if_over(5, take_first)) %>%
  # Empty the ancestor of animals and fungi 
  clear_uncle(2759, 33154) %>%
  # Replace the representatives with humans and yeast
  add_representative(scinames=c('Homo sapiens', 'Saccharomyces cerevisiae')) %>%
  # Use a prebuilt set of prokaryotic species
  use_recommended_prokaryotes

  # TODO: add archaeplastida (viridiplantae, red algae, glaucophytes)
```

```{r, eval=FALSE}
# TODO: wrap this up neatly internally yadayada

# add the focal species to the blast set
strata[[focal_taxid]][[focal_taxid]] <- focal_taxid

# Download genomes, storing the filenames
strata_fasta <- uniprot_fill_strata(strata)

# extract the FASTA file for the focal species
query <- strata_fasta[[focal_taxid]][[focal_taxid]]

# BLAST against each target genome (this will take a few hours)
# TODO: don't hard code the threads
results <- strata_blast(query, strata_fasta, blast_args=list(nthreads=8)) %>%
  strata_besthits %>%
  merge_besthits
```

The `results` variable is saved as a `phylostratr` dataset, and can be loaded
as below:

```{r, eval=FALSE}
data(arabidopsis)
head(arabidopsis)
```

To get gene list according to phylostrata:

```{r, eval=FALSE}
# TODO: add locus mrca
phylostrata <- stratify(arabidopsis)
head(phylostrata)
```

And get the number of genes in each phylostrata

```{r, eval=FALSE}
summary(phylostrata$mrca_name)
```

Note that this is a reduced dataset, with only 1/100 of the genes represented.
