---
title: "*Saccharomyces cerevisiae* case study"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{*Saccharomyces cerevisiae* case study}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

The main practical difference between this vignette and the Arabidopsis
vignette is that for *S. cerevisiae* the NCBI common tree is not high enough
resolution. The *Saccharomyces* genera includes many species disant species, so
we cannot just find the NCBI taxonomy backbone nodes and then download the
needed sequences from uniprot (indeed, UniProt doesn't even have all the
species we want).

The most general way to customize a sequence dataset is to simply write the
FASTA filenames into a list and convert it to a tree:

```{r}
saccharomyces <- list(
  list(
    name='s6',
    list(name='Saccharomyces_eubayanus', seq='yeast/eubayanus.faa'),
    list(name='Saccharomyces_uvarum', prot='yeast/uvarum.faa'),
    list(
      name='s5',
      list(name='Saccharomyces_arboricola', prot='yeast/arboricola.faa'),
      list(
        name='s4',
        list(name='Saccharomyces_kudriavzevii', prot='yeast/kudriavzevii.faa'),
        list(
          name='s3',
          list(name='Saccharomyces_mikatae', prot='yeast/mikatae.faa'),
          list(
            name='s2',
            list(name='Saccharomyces_paradoxus', prot='yeast/paradoxus.faa'),
            list(
              name='s1',
              list(name='Saccharomyces_cerevisiae', prot='yeast/cerevisiae.faa')
            )
          )
        )
      )
    )
  )
) %>% data.tree::FromListSimple()
saccharomyces
```

Where we assume the folder `yeast` holds the respective protein sequences. The
phylostrata names (s1-6) are arbitrary. It is important to include the focal
species, cerevisiae. This tree could also be loaded from a YAML, JSON, or
NEWICK file. For more information on loading trees, see the documentation for
the `data.tree` R package.

To get deeper phylostrata, we will use the UniProt genomes and the NCBI taxonomy
tree. As in the *Arabidopsis* vignette, we don't want to use all the UniProt
sequences, so we will filter for diverse species. 

```{r, eval=FALSE}
# Get all UniProt proteomes
deep_strata <- uniprot_cousins(4932) %>%
  # Make filter, if there are more than 5 species in a stratum, take just the
  # first species in each uncle. I am not really happy with this solution ...
  # TODO: add filter for reference species
  lapply(make_do_if_over(5, take_first)) %>%
  # add prokaryote stratum
  use_recommended_prokaryotes %>%
  # download UniProt sequences (this may take 10+ minutes)
  uniprot_fill_strata %>%
  # replace taxon ids with taxon names
  as_named_strata(depth=2)
```

Now we will use the *Saccharomyces* tree we defined above to resolve
phylogenetic relationships among the species in the *Saccharomyces* genus,
this allows for finer phylostratigraphic analysis.

```{r, eval=FALSE}
# remove the youngest strata (Saccharomyces)
strata <- append(head(deep_strata, -1), saccharomyces)
```

Next we run BLAST

```{r, eval=FALSE}
results <- strata_blast(
    query      = saccharomyces$s1[[1]],
    strata     = strata,
    blast_args = list(nthreads=8) # TODO: don't hardcode the threads
) %>%
  strata_besthits %>%
  merge_besthits
```
