---
title: "*Saccharomyces cerevisiae* case study"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{*Saccharomyces cerevisiae* case study}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

The main practical difference between this vignette and the Arabidopsis
vignette is that for *S. cerevisiae* the NCBI common tree is not high enough
resolution. The *Saccharomyces* genera includes many species disant species, so
we cannot just find the NCBI taxonomy backbone nodes and then download the
needed sequences from uniprot (indeed, UniProt doesn't even have all the
species we want).

The most general way to customize a sequence dataset is to simply write the
FASTA filenames into a list and convert it to a tree:

```{r, eval=FALSE}
library(phylostratr)
library(magrittr)
focal_taxid <- 4932
saccharomyces <- Strata(
    tree = ape::read.tree(system.file('extdata', 'yeast', 'tree', package='phylostratr')),
    data = list(faa=list(
             Saccharomyces_cerevisiae   = 'yeast/cerevisiae.faa',
             Saccharomyces_paradoxus    = 'yeast/paradoxus.faa',
             Saccharomyces_mikatae      = 'yeast/mikatae.faa',
             Saccharomyces_kudriavzevii = 'yeast/kudriavzevii.faa',
             Saccharomyces_arboricola   = 'yeast/arboricola.faa',
             Saccharomyces_eubayanus    = 'yeast/eubayanus.faa',
             Saccharomyces_uvarum       = 'yeast/uvarum.faa'
           )),
    focal_species = 'Saccharomyces_cerevisiae'
) %>% strata_convert(target='tip', to='id')
```

Where we assume the folder `yeast` holds the respective protein sequences. It
is important to include the focal species, cerevisiae. This tree could also be
loaded from a YAML, JSON, or NEWICK file. For more information on loading
trees, see the documentation for the `data.tree` R package.

To get deeper phylostrata, we will use the UniProt genomes and the NCBI taxonomy
tree. As in the *Arabidopsis* vignette, we don't want to use all the UniProt
sequences, so we will filter for diverse species. 

```{r, eval=FALSE}
# set weights on species selection
weights <- c(
  '1355160' = 0,   # these three are deprecated
  '245562'  = 0,
  '1094981' = 0,
  '284813'  = 1.1, # these are the reference replacements
  '766039'  = 1.1,
  '237561'  = 1.1
)
focal_taxid <- '4932'
strata <- focal_taxid %>%
  # Get all UniProt proteomes
  uniprot_strata %>%
  # build a tree of all UniProt genomes
  # Select a diverse subset of 5 or fewer representatives from each stratum.
  # Only do this above the Saccharomyces genus, since we will later replace
  # Saccharomyces with out own tree.
  strata_apply(f=diverse_subtree, n=5, weights=weights) %>%
  # add prokaryote stratum
  use_recommended_prokaryotes %>%
  # download UniProt sequences (this may take 10+ minutes)
  uniprot_fill_strata
```

Now we will use the *Saccharomyces* tree we defined above to resolve
phylogenetic relationships among the species in the *Saccharomyces* genus,
this allows for finer phylostratigraphic analysis.

```{r, eval=FALSE}
# Remove the Saccharomyces branch
strata@tree <- prune(strata@tree, '4930', type='name')
strata <- add_taxid(strata, '4930')
# Replace it with out custom tree (4892 is the parent of 4930)
strata@tree <- ape::bind.tree(strata@tree, saccharomyces@tree, where=which(tree_names(strata@tree) == '4930'))
strata@data$faa <- append(strata@data$faa, saccharomyces@data$faa)
```

Next we run BLAST

```{r, eval=FALSE}
results <- strata_blast(
    strata     = strata,
    blast_args = list(nthreads=8)
) %>%
  strata_besthits %>%
  merge_besthits

pdf('all-yeast.pdf')
plot_obo_trees(results, tree=strata@tree, focal_id=focal_taxid)
dev.off()
```
